File: CS/lab2.3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CS23
{
    public class Book
    {
        public string ISBN { get; set; }
        public string Name { get; set; }
        public List<string> Authors { get; set; } = new List<string>();
        public string Publish { get; set; }
        public string Genre { get; set; }
        public string Date { get; set; }

        public void AddAuthor(string Author)
        {
            Authors.Add(Author);
        }
        public override string ToString()
        {
            return $"Book \nISBN: '{ISBN}',Name: '{Name}', Authors: '{string.Join(", ", Authors)}',Publish: '{Publish}', Genre: '{Genre}', Date:'{Date}'";
        }
    }

    class BookContainer
    {
        private List<Book> books;

        public BookContainer(List<Book> books)
        {
            this.books = books;
        }

        public List<Book> SortByName()
        {
            List<Book> sortedBooks = new List<Book>(books);
            sortedBooks.Sort((b1, b2) => string.Compare(b1.Name, b2.Name, StringComparison.Ordinal));
            return sortedBooks;
        }

        public List<Book> SortByPublish()
        {
            List<Book> sortedBooks = new List<Book>(books);
            sortedBooks.Sort((b1, b2) => string.Compare(b1.Publish, b2.Publish, StringComparison.Ordinal));
            return sortedBooks;
        }

        public List<Book> SortByGenre()
        {
            List<Book> sortedBooks = new List<Book>(books);
            sortedBooks.Sort((b1, b2) => string.Compare(b1.Genre, b2.Genre, StringComparison.Ordinal));
            return sortedBooks;
        }

        public List<Book> SortByDate()
        {
            List<Book> sortedBooks = new List<Book>(books);
            sortedBooks.Sort((b1, b2) => DateTime.Compare(DateTime.Parse(b1.Date), DateTime.Parse(b2.Date)));
            return sortedBooks;
        }
    }

    internal class lab23
    {
        static void Main(string[] args)
        {
            List<Book> books = new List<Book>();
            books.Add(new Book { ISBN = "978142890", Name = "The Adventure Chronicles", Authors = new List<string> { "John Smith" }, Publish = "Adventure Books Inc", Genre = "Adventure", Date = "2022-08-15" });
            books.Add(new Book { ISBN = "123456789", Name = "The Mysterious Island", Authors = new List<string> { "Sarah Brown", "Alex Johnson" }, Publish = "Mystery Novels Publishing", Genre = "Mystery", Date = "2023-04-10" });
            books.Add(new Book { ISBN = "567890123", Name = "Journey to the Center of the Earth", Authors = new List<string> { "George Orwell" }, Publish = "Exploration Publications", Genre = "Adventure", Date = "2021-12-05" });
            books.Add(new Book { ISBN = "987654321", Name = "The Code Breaker", Authors = new List<string> { "Walter Isaacson" }, Publish = "Simon & Schuster", Genre = "Non-fiction", Date = "2023-02-20" });
            books.Add(new Book { ISBN = "111223344", Name = "To Kill a Mockingbird", Authors = new List<string> { "Harper Lee" }, Publish = "J.B. Lippincott & Co.", Genre = "Fiction", Date = "2021-07-11" });

            BookContainer bookContainer = new BookContainer(books);

            // Створення завдань для паралельної обробки
            var tasks = new List<Task<List<Book>>>();
            tasks.Add(Task.Run(() => bookContainer.SortByName()));
            tasks.Add(Task.Run(() => bookContainer.SortByPublish()));
            tasks.Add(Task.Run(() => bookContainer.SortByGenre()));
            tasks.Add(Task.Run(() => bookContainer.SortByDate()));

            // Очікування завершення всіх завдань
            Task.WhenAll(tasks).Wait();

            // Отримання результатів паралельної обробки
            List<Book> resultByName = tasks[0].Result;
            List<Book> resultByPublish = tasks[1].Result;
            List<Book> resultByGenre = tasks[2].Result;
            List<Book> resultByDate = tasks[3].Result;

            // Порівняння часу послідовної обробки
            var stopwatch = new System.Diagnostics.Stopwatch();
            stopwatch.Start();
            List<Book> sequentialResultByName = bookContainer.SortByName();
            List<Book> sequentialResultByPublish = bookContainer.SortByPublish();
            List<Book> sequentialResultByGenre = bookContainer.SortByGenre();
            List<Book> sequentialResultByDate = bookContainer.SortByDate();
            stopwatch.Stop();
            long sequentialTime = stopwatch.ElapsedTicks;

            Console.WriteLine("Sequential processing time: " + sequentialTime + " ticks");

            // Обробка і виведення результатів паралельної обробки
            Console.WriteLine("\nSort results by name (in parallel):");
            foreach (Book book in resultByName)
            {
                Console.WriteLine(book.Name);
            }

            Console.WriteLine("\nSort results by publish (in parallel):");
            foreach (Book book in resultByPublish)
            {
                Console.WriteLine(book.Publish);
            }

            Console.WriteLine("\nSort results by genres (in parallel):");
            foreach (Book book in resultByGenre)
            {
                Console.WriteLine(book.Genre);
            }

            Console.WriteLine("\nSort results by date (in parallel):");
            foreach (Book book in resultByDate)
            {
                Console.WriteLine(book.Date);
            }

            Console.ReadKey();
        }
    }
}

File: Java/src/lab23.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

class Book {
    private String ISBN;
    private String Name;
    private List<String> Authors;
    private String Publish;
    private String Genre;
    private String Date;
    public Book(String ISBN, String Name, String Publish, String Genre, String Date){
        this.ISBN = ISBN;
        this.Name = Name;
        this.Authors = new ArrayList<>();
        this.Publish = Publish;
        this.Genre = Genre;
        this.Date = Date;
    }
    public void AddAuthor(String Author) {
        Authors.add(Author);
    }
    public String getISBN(){
        return this.ISBN;
    }
    public String getName(){
        return this.Name;
    }
    public String getPublish(){
        return this.Publish;
    }
    public String getGenre(){
        return this.Genre;
    }
    public String getDate(){
        return this.Date;
    }
}

class BookContainer {
    private List<Book> books;
    public BookContainer(List<Book> books)
    {
        this.books = books;
    }

    public synchronized List<Book> SortByName() {
        List<Book> sortedRoutes = new ArrayList<>(books);
        Collections.sort(sortedRoutes, (r1, r2) -> r1.getName().compareTo(r2.getName()));
        return sortedRoutes;
    }

    public synchronized List<Book> SortByPublish() {
        List<Book> sortedRoutes = new ArrayList<>(books);
        Collections.sort(sortedRoutes, (r1, r2) -> r1.getPublish().compareTo(r2.getPublish()));
        return sortedRoutes;
    }

    public synchronized List<Book> SortByGenre() {
        List<Book> sortedRoutes = new ArrayList<>(books);
        Collections.sort(sortedRoutes, (r1, r2) -> r1.getGenre().compareTo(r2.getGenre()));
        return sortedRoutes;
    }

    public synchronized List<Book> SortByDate() {
        List<Book> sortedRoutes = new ArrayList<>(books);
        Collections.sort(sortedRoutes, (r1, r2) -> r1.getDate().compareTo(r2.getDate()));
        return sortedRoutes;
    }
}
public class lab23 {
    public static void main(String[] args) {
        List<Book> books = new ArrayList<>();

        books.add(new Book("978142890", "The Mysterious Island", "John Smith", "AdventureInc", "2022-08-15"));
        books.add(new Book("123456789", "Journey to the Center of the Earth", "Mystery Novels Publishing", "Mystery", "2023-04-10"));
        books.add(new Book("567890123", "The Code Breaker", "Exploration Publications", "Adventure", "2021-12-05"));
        books.add(new Book("987654321", "To Kill a Mockingbird", "Simon & Schuster", "Non-fiction", "2023-02-20"));
        books.add(new Book("111223344", "The Adventure Chronicles", "J.B. Lippincott & Co.", "Fiction", "2021-07-11"));

        BookContainer container = new BookContainer(books);

        ExecutorService executorService = Executors.newFixedThreadPool(4);

        List<Book> resultByName = new ArrayList<>();
        List<Book> resultByPublish = new ArrayList<>();
        List<Book> resultByGenre = new ArrayList();
        List<Book> resultByDate = new ArrayList();

        executorService.execute(() -> {
            resultByName.addAll(container.SortByName());
        });

        executorService.execute(() -> {
            resultByPublish.addAll(container.SortByPublish());
        });

        executorService.execute(() -> {
            resultByGenre.addAll(container.SortByGenre());
        });

        executorService.execute(() -> {
            resultByDate.addAll(container.SortByDate());
        });

        executorService.shutdown();

        try {
            executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        long startTime = System.nanoTime();
        List<Book> sequentialResultByName = container.SortByName();
        List<Book> sequentialResultByPublish = container.SortByPublish();
        List<Book> sequentialResultByGenre = container.SortByGenre();
        List<Book> sequentialResultByDate = container.SortByDate();
        long endTime = System.nanoTime();
        long sequentialTime = endTime - startTime;

        System.out.println("Sequential processing time: " + sequentialTime + " nanosecond");

        System.out.println("\nSort results by name (in parallel):");
        for (Book book : resultByName) {
            System.out.println(book.getName());
        }

        System.out.println("\nSort results by publish (in parallel):");
        for (Book book : resultByPublish) {
            System.out.println(book.getPublish());
        }

        System.out.println("\nSort results by genre (in parallel):");
        for (Book book : resultByGenre) {
            System.out.println(book.getGenre());
        }

        System.out.println("\nSort results by date (in parallel):");
        for (Book book : resultByDate) {
            System.out.println(book.getDate());
        }
    }
}